### PR TITLE
[14.0][MIG] sale_promotion_rule_assortment

### PR DESCRIPTION
**Pull request by xavier-bouquiaux**

### CODE REVIEW

**Comment by rousseldenis** [comment]
Depends on:

- [ ] #1583

**Comment by elvise** [comment]
hi @sbidoul can you check the runboat ?

**Comment by sbidoul** [comment]
@xavier-bouquiaux can you rebase this PR ?

**Comment by xavier-bouquiaux** [comment]
> @xavier-bouquiaux can you rebase this PR ?

Done

**Comment by xavier-bouquiaux** [comment]
rebased as #1583 has been merged

**Review by rousseldenis — APPROVED** [review]

**Comment by xavier-bouquiaux** [comment]
@acsonefho

File: sale_promotion_rule_assortment/models/sale_promotion_rule.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            return product_obj.search(domain)
+        return product_obj.search([])
+
+    def _get_promotions_valid_order_lines(self, order=False, line=False):
+        if order:
+            order_lines = order.order_line
+        elif line:
+            order_lines = line
+        if not self.filter_id:
+            return order_lines
```
Comment (acsonefho) [inline sale_promotion_rule_assortment/models/sale_promotion_rule.py]: If no `order` and no `line`, you will return `order_lines` before declaring it.

File: sale_promotion_rule_assortment/models/sale_promotion_rule.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        promotion_product_ids = self._get_promotion_rule_products()
+        order_line_ids = order_lines.filtered(
```
Comment (acsonefho) [inline sale_promotion_rule_assortment/models/sale_promotion_rule.py]: Same here: `order_lines` may not exist.

File: sale_promotion_rule_assortment/models/sale_promotion_rule.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            lambda line, product_ids=promotion_product_ids: line.product_id.id
+            in product_ids.ids
+        )
+        return order_line_ids
+
+    @api.model
+    def _get_restrictions(self):
+        res = super(SalePromotionRule, self)._get_restrictions()
```
Comment (acsonefho) [inline sale_promotion_rule_assortment/models/sale_promotion_rule.py]: Please use python3 format: `super()._get_restrictions()`

File: sale_promotion_rule_assortment/models/sale_promotion_rule.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    def _check_valid_product_assortment(self, order):
+        if self.filter_id:
+            order_lines = self._get_promotions_valid_order_lines(order=order)
+            if not order_lines:
+                return False
+
+        return True
+
+    def _is_promotion_valid_for_line(self, line):
+        res = super(SalePromotionRule, self)._is_promotion_valid_for_line(line)
```
Comment (acsonefho) [inline sale_promotion_rule_assortment/models/sale_promotion_rule.py]: Same here: `super()._is_promotion_valid_for_line(line)`

File: sale_promotion_rule_assortment/tests/test_promotion_rule_assortment.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+from odoo.tests.common import TransactionCase
+
+from odoo.addons.sale_promotion_rule.tests.test_promotion import (
+    AbstractCommonPromotionCase,
+)
+
+
+class TestPromotionRuleAssortment(TransactionCase, AbstractCommonPromotionCase):
+    def setUp(self):
+        super(TestPromotionRuleAssortment, self).setUp()
```
Comment (acsonefho) [inline sale_promotion_rule_assortment/tests/test_promotion_rule_assortment.py]: `super().setUp()`

**Review by acsonefho — CHANGES_REQUESTED** [review]

**Comment by xavier-bouquiaux** [comment]
@rousseldenis : this one is mergeable i think

**Review by acsonefho — APPROVED** [review]

**Comment by rousseldenis** [comment]
@xavier-bouquiaux Could you squash those commits ? 

image

**Comment by acsonefho** [comment]
> @xavier-bouquiaux Could you squash those commits ?
> 
> image

@xavier-bouquiaux Can you squash please :pray: Then @rousseldenis can merge it :muscle:

**Comment by xavier-bouquiaux** [comment]
@rousseldenis @acsonefho : squash done  :can we merge ?

**Pull request merged**